---
#title: "Subset analyses"
output: md_document
params:
  input_file: "~/FRCBS/interval_prediction/data/full_data_preprocessed-2020-05-25-train.rdata"
  #input_file: "~/preprocessed_data_6243aede7.rdata"
  hlen: NULL
  hlen_exactly: FALSE
  sample_fraction: 1.0
  method: "both"
  gender: "both"
  id: NULL
  # sample_fraction: 0.004
  # method: "icp-fix"
  # gender: "male"
  # id:   date-2020-07-17-gender-male-method-icp-fix-sample_fraction-0.004
  date: "2020-07-08"
  extra_id: NULL
  errors_table_file: NULL
  Hb_cutoff_male: 135
  Hb_cutoff_female: 125
  cores: 4
  iterations: 2000
  skip_train: FALSE
  create_datasets_bool: TRUE
#  donor_specific_file: "~/FRCBS/interval_prediction/data/finngen_snip_prs-2020-07-29.rdata"
  donor_specific_file: NULL
---

  
```{r, results="asis", echo=FALSE}
cat(sprintf("# Subset analyses (%s)\n", params$gender))
```

<!--
<h2> Introduction </h2>

Here I will run analyses for three different models with varying parameters. Each model is run with and without icp-fix and separately for male and female donors resulting in 12 analyses in total. The models are following:

*1.* Linear mixed effect model with random intercept for each donor. Stan-code is in file 'mem_qr_1.stan':

\[
y_{it} = \alpha + b_i + \beta' x_{it} + \epsilon
\]

where $\epsilon \sim N(0, \sigma_{\epsilon}^2), \quad b_i \sim N(0,\sigma_{b}^2)$

*2.* Linear mixed effect model with random intercept for each donor and a random slope for age variable:

\[
y_{it} = \alpha + b_{0i} + (b_{1i} + \beta_0)x_{age,it} + \beta_1' x_{1it} + \epsilon
\]

where $\epsilon \sim N(0, \sigma_{\epsilon}^2), \quad b_{0i}, b_{1i} \sim N(0,\sigma_{b}^2)$

*3.* Linear mixed effect model with random intercept for each donor and a random slope for days_to_previous_fb variable:

\[
y_{it} = \alpha + b_{0i} + (b_{1i} + \beta_0)x_{dtp,it} + \beta_1' x_{1it} + \epsilon
\]

where $\epsilon \sim N(0, \sigma_{\epsilon}^2), \quad b_{0i}, b_{1i} \sim N(0,\sigma_{b}^2)$

The Wooldridge intial conditions problem fix replaces the individual random intercept term $b_i$ with the following term:

\[
b_i = \zeta_0 + \zeta_1 y_{0i} + \upsilon' z_{0i} + a_i
\]

where $a_i \sim N(0,\sigma_a^2)$ and $z$ are exogenous variables that could be associated with the initial observation. Some other variants of Wooldridge solution use the mean of exogenous variables or associate means of other covariates into the model. In this solution $a_i$ is our new random intercept term which is assumed to be uncorrelated with the initial observations $y_{0i}$ satisfying the model assumptions.

The goal of this document is to get a feeling which model seems to work best for our data. This can be used to decide which models should be run on the whole dataset. We can see does the slope variable in models 2 and 3 improve the predictions and if the ICP-fix is suitable for the analyses.

-->

```{r, echo=FALSE}
suppressPackageStartupMessages(library(tidyverse))
```

```{r, echo = FALSE}

# Variables that are in use

descript <- tibble(Variable = c("donor", "Hb", "days_to_previous_fb", "age", "previous_Hb_def", 
                                "year", "warm_season", "consecutive_deferrals", "recent_donations",
                                "recent_deferrals", "hour", 
                                "previous_Hb", "Hb_first", "Hb_deferral"), 
                   Pretty = c("Donor ID", "Hemoglobin", "Days to previous full blood donation", "Age", "Previous Hb deferral", 
                              "Year", "Warm season", "Consecutive deferrals", "Recent donations", 
                              "Recent deferrals", "Hour", 
                              "Previous Hb", "First Hb", "Hb deferral"),
                   Type = c("Factor", "numeric", "numeric (int)", "numeric", "boolean",
                            "numeric (int)", "boolean", "numeric (int)", "numeric (int)", "numeric (int)", "numeric",
                            "numeric", "numeric", "boolean"),
                   Explanation = c("Donor identifier",
                                   "Amount of Hemoglobin",
                                   "Time (in days) between Hb measurement and previous full blood donation event",
                                   "Age of donor",
                                   "Indicates whether the donor was deferred from blood donation due to low hemoglobin at previous donation event",
                                   "Year of donation",
                                   "True if donation was given in April-September",
                                   "Number of times the donor has been deferred due to low hemoglobin since last succesful whole blood donation",
                                   "Number of donations in the last two years",
                                   "Number of deferrals due to low hemoglobin in the last two years",
                                   "Time of day when donation was given as hours (e.g. 13:45 = 13.75)",
                                   "Hb value at previous measurement (ICP-model)",
                                   "Hb value at first donation of this donor (Non ICP-model)",
                                   "Deferred based on low hemoglogin")
                   )

donor_descript <- tibble(
  Variable    = c("smoking", "height", "weight", "RNF43_mutant", "prs"),
  Pretty      = c("Smoking", "Height", "Weight", "RNF43 mutant", "Polygenic risk score"),
  Type        = c("boolean", "numeric", "numeric", "boolean", "numeric"),
  Explanation = c("Does the person smoke", "Height of the donor", "Weight of the donor", 
                   "Mutation at RNF43 gene in chromosome 17 position 58358769", "Polygenic risk score for hemoglobin")
)
  
to_pretty <- function(df, description) {
  old_names <- colnames(df)
  new_names <- description %>% filter(Variable %in% all_of(old_names)) %>% pull(Pretty)
  colnames(df) <- new_names
  df
}

```

<!-- <h2> Initialization </h2> -->
```{r Setup, setup = TRUE, echo=FALSE, message=FALSE}

suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(knitr))
library(rstan)
suppressPackageStartupMessages(library(caret))
suppressPackageStartupMessages(library(tictoc))
suppressPackageStartupMessages(library(entropy))
suppressPackageStartupMessages(library(brms))
suppressPackageStartupMessages(library(ggmcmc))
suppressPackageStartupMessages(library(ggthemes))
suppressPackageStartupMessages(library(ggridges))
suppressPackageStartupMessages(library(ROCR))
suppressPackageStartupMessages(library(pROC))
suppressPackageStartupMessages(library(furrr))
suppressPackageStartupMessages(library(sn))
library(tidyselect)

knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE)
#knitr::opts_chunk$set(echo=TRUE, message=TRUE)
#options(warn = 1)


message("Parameters are:")
for (name in names(params)) {
  message(sprintf("%s = ", name), params[[name]])
}

if (knitr::opts_chunk$get("dev") == "pdf") {
  knitr::opts_chunk$set(dev="cairo_pdf")
  knitr::opts_current$set(dev="cairo_pdf")
}
message(paste0("Device is ", knitr::opts_chunk$get("dev"), "\n"))
message(paste0("Current device is ", knitr::opts_current$get("dev"), "\n"))
use_pdf <- knitr::opts_chunk$get("dev") %in% c("pdf", "cairo_pdf")

# These are used to select which computations to perform,
# base on params$gender and params$method
compute_male   <- params$gender %in% c("male", "both")
compute_female <- params$gender %in% c("female", "both")
compute_nofix  <- params$method %in% c("no-fix", "both")
compute_icpfix <- params$method %in% c("icp-fix", "both")
compute_male_nofix <- compute_male && compute_nofix 
compute_female_nofix <- compute_female && compute_nofix 
compute_male_icpfix <- compute_male && compute_icpfix 
compute_female_icpfix <- compute_female && compute_icpfix 

set.seed(123)
if (!is.null(params$cores)) {
  options(mc.cores = params$cores)
}

# Set variables to save time knitting the document
#skip_train = params$skip_train
skip_train = use_pdf
#create_datasets_bool = params$create_datasets_bool
create_datasets_bool = ! use_pdf
knit = TRUE

datadir = "../data/rdata/"
dumpdir = "../data/rdump/"
stanfitpath = "../data/stan_fits/"
rawresultdir = "../data/raw_results/"

if (is.null(params$id)) {
  id <- paste(params$date, params$sample_fraction, params$hlen, params$hlen_exactly, params$extra_id, sep="_") 
} else {
  id <- params$id
}
data_id <- id

fit_id = id


source("helper_functions.R")
source("validate_stan_fit.R")
#source("ppc.R")
```

```{r Load datas, echo=FALSE, message=FALSE}

data <- load_single(params$input_file) # Forgets the name of the object

data <- data %>%
  mutate(days_to_previous_fb = as.double(days_to_previous_fb),
         donor = as.character(donor))   # For some reason this cannot be a factor

#variables_to_log <- c("days_to_previous_fb",
#                                "consecutive_deferrals", "recent_donations",
#                                "recent_deferrals")

#data <- data %>%
#  mutate_at(variables_to_log, function(x) log(x+1))

sample_fraction <- params$sample_fraction

is.wholenumber <- function(x, tol = .Machine$double.eps^0.5)  abs(x - round(x)) < tol

if (sample_fraction > 1 && is.wholenumber(sample_fraction)) {  # If instead of fraction a sample size is given, convert it to sample fraction
  low_limit <- ifelse(params$method=="no-fix", 2, 1)
  data <- data %>% group_by(donor) %>%
    filter(n() > low_limit) %>%
    ungroup()
  
  if (params$gender == "male") {
    temp <- data %>% filter(gender=="Men")
  }
  else if (params$gender == "female") {
    temp <- data %>% filter(gender=="Women")
  } else {
    temp <- data
  }
  n <- length(unique(temp$donor))  # number of donors
  rm(temp)
  sample_fraction <- sample_fraction / n
  message(sprintf("Number of donors: %i, sample_fraction: %f", n, sample_fraction))
}

if (!is.null(params$donor_specific_file) && (length(params$donor_specific_file) > 0)) {
  #donor_variables <- c("smoking", "bmi", "RNF43_mutant", "prs")
  #donor_variables <- c("smoking", "height", "weight", "RNF43_mutant", "prs")
  #pretty_donor_variables <- c("Smoking", "Height", "Weight", "RNF43 mutant", "Polygenic risk score")
  donor_variables <- donor_descript$Variable
  pretty_donor_variables <- donor_descript$Pretty
  
  findonor_and_snip_and_prs <- load_single(params$donor_specific_file)
  data <- inner_join(data, findonor_and_snip_and_prs, by="donor")
} else {
  donor_variables = NULL
}

# Stan parameter names
pars1 <- sprintf("beta[%i]", 1:10)
pars1.icp <- c(pars1,
               sprintf("ups[%i]", 1:4))
if (!is.null(donor_variables)) {
  donor_pars <- sprintf("phi[%i]", 1:length(donor_variables))
  pars1 <- c(pars1, donor_pars)
  pars1.icp <- c(pars1.icp, donor_pars)
}

# Pretty names for parameters
vars1 <- descript %>% filter(!Variable %in% c("donor", "Hb", "previous_Hb", "Hb_deferral")) %>% pull(Pretty)   # Convert to pretty names

vars1.icp <- descript %>% filter(!Variable %in% c("donor", "Hb", "Hb_first", "Hb_deferral")) %>% pull(Pretty)   # Convert to pretty names
vars1.icp <- c(vars1.icp,                    
                   c("First age", "First year", "First warm season", "First hour"))
if (!is.null(donor_variables)) {
  vars1     <- c(vars1, pretty_donor_variables)
  vars1.icp <- c(vars1.icp, pretty_donor_variables)
}

# Make sure the columns in data are in the same order as in the vars and pars lists
variables_in_order <- setdiff(descript$Variable, "donor")
if (!is.null(donor_variables)) {
  data <- data %>% select(all_of(variables_in_order),
                          all_of(donor_variables), everything())
} else {
  data <- data %>% select(all_of(variables_in_order), everything())
}

if (create_datasets_bool) {
  source("jarkko_create_datasets.R")
  lst <- create_datasets(data, datadir, dumpdir, data_id, sample_fraction, hlen=params$hlen, params$hlen_exactly,
                         Hb_cutoff_male = params$Hb_cutoff_male,
                         Hb_cutoff_female = params$Hb_cutoff_female, donor_variables=donor_variables,
                         compute_male_nofix, compute_female_nofix, compute_male_icpfix, compute_female_icpfix)
  
  data.male <- lst$small.data.male
  data.female <- lst$small.data.female
} else {   # This is only for plotting purpose
  data.male <-
    data %>% filter(gender == "Men")
  data.female <-
    data %>% filter(gender == "Women")
  # Split for men and women separately
  data.male <- split_set(data.male, 1.0)
  data.female <- split_set(data.female, 1.0)
  names(data.male) <- c("general","validation")
  names(data.female) <- c("general","validation")
  # Small test sets
  smallm <- split_set(data.male$general, sample_fraction)
  smallf <- split_set(data.female$general, sample_fraction)
  data.male <- smallm$train
  data.female <- smallf$train
}

# This in only for plotting purpose
data.male <- filter_based_on_number_of_donations(data.male, params$hlen, params$hlen_exactly)
data.female <- filter_based_on_number_of_donations(data.female, params$hlen, params$hlen_exactly)


if (is.null(donor_variables)) {
  stanf1     <- "stan/container.stan"
  stanf1.icp <- "stan/container_heckman.stan"
} else {
  stanf1     <- "stan/container_consts.stan"
  stanf1.icp <- "stan/container_heckman_consts.stan"
}


error_rows <- list()  # This is used to construct the error table in the end
                   # It has columns Model, MAE, RMSE

```



## Data description



### Variables used in prediction

```{r, echo=FALSE}
kable(descript)
kable(donor_descript)
```


```{asis, echo=compute_male}
### Summary plots of variables (male)
```

```{r Summary plots for males, echo = FALSE, eval=compute_male}
#Plot distributions for male donors


data.male2 <- data.male %>%
  filter(first_event == FALSE) %>%
#         !is.na(Hb_first), !is.na(days_to_previous_fb),
#         !is.na(previous_Hb), 
#         donat_phleb == 'K' | donat_phleb == '*' | donat_phleb == 'T',
#         !is.na(previous_Hb_def)) %>%
  select(all_of(descript$Variable)) %>%
  mutate(Hb_deferral = as.integer(Hb_deferral),
         warm_season = as.integer(warm_season),
         previous_Hb_def = as.integer(previous_Hb_def))

pmale <- data.male2 %>%
  keep(is.numeric) %>%
  to_pretty(descript) %>%
  gather() %>%
  mutate(key = factor(key, levels=descript$Pretty))  %>% # Don't sort alphabetically
  ggplot(aes(value)) +
  facet_wrap(~ key, scales = "free") +
  geom_histogram(fill = "blue")
pmale

if (!is.null(params$donor_specific_file)) {
  male_donor_specific <- findonor_and_snip_and_prs %>%
    semi_join(data.male2, by="donor") %>%
    mutate(smoking = as.integer(smoking), RNF43_mutant = as.integer(RNF43_mutant)) %>%
    keep(is.numeric) %>%
    to_pretty(donor_descript) 
  pmale2 <- male_donor_specific %>%
    gather() %>%
    mutate(key = factor(key, levels=donor_descript$Pretty))  %>% # Don't sort alphabetically
    ggplot(aes(value)) +
    facet_wrap(~ key, scales = "free") +
    geom_histogram(fill = "blue")
  pmale2
}
rm(data.male2);
```



```{asis, echo=compute_female}
### Summary plots of variables (female)
```

```{r Summary plots for females, echo = FALSE, eval=compute_female}
#Plot distributions for female donors

data.female2 <- data.female %>%
  filter(first_event == FALSE) %>%
#         !is.na(Hb_first), !is.na(days_to_previous_fb),
#         !is.na(previous_Hb), 
#         donat_phleb == 'K' | donat_phleb == '*' | donat_phleb == 'T',
#         !is.na(previous_Hb_def)) %>%
  select(all_of(descript$Variable)) %>%
  mutate(Hb_deferral = as.integer(Hb_deferral),
         warm_season = as.integer(warm_season),
         previous_Hb_def = as.integer(previous_Hb_def))

pfemale <- data.female2 %>%
          keep(is.numeric) %>%
          to_pretty(descript) %>%
          gather() %>%
          mutate(key = factor(key, levels=descript$Pretty))  %>% # Don't sort alphabetically
          ggplot(aes(value)) +
            facet_wrap(~ key, scales = "free") +
            geom_histogram(fill = "orange")

pfemale

if (!is.null(params$donor_specific_file)) {
  female_donor_specific <- findonor_and_snip_and_prs %>%
    semi_join(data.female2, by="donor") %>%
    mutate(smoking = as.integer(smoking), RNF43_mutant = as.integer(RNF43_mutant))
  pfemale2 <- female_donor_specific %>%
    keep(is.numeric) %>%
    to_pretty(donor_descript) %>%
    gather() %>%
    mutate(key = factor(key, levels=donor_descript$Pretty))  %>% # Don't sort alphabetically
    ggplot(aes(value)) +
    facet_wrap(~ key, scales = "free") +
    geom_histogram(fill = "orange")
  pfemale2
}
rm(data.female2);
```


```{r clean up memory, echo = FALSE}
rm(data); invisible(gc())
rm(data.male, data.female); invisible(gc(reset = TRUE))
```





## Results


```{asis, echo=compute_male_nofix}
### Model 1, male donors
```

```{r Model 1 male donors, results="hide", eval=compute_male_nofix}
# Load first datasets
message("0")
load(paste(datadir,"stan_preprocessed_datasets_",data_id,".RData", sep = '')) 
load(paste(datadir,"stan_lists_male_nofix_",data_id,".RData", sep = ''))       
cat("1\n")
message("1")
if (skip_train == TRUE) {
  load(paste(stanfitpath, "small_model1_male_", fit_id, ".RData", sep = ''))
    } else {
  fit <- stan(file = stanf1, 
            data = stan.lists.male.nofix,
            save_warmup = TRUE,
            iter = params$iterations,
            control = list(max_treedepth = 12))
  save(fit, file = paste(stanfitpath, "small_model1_male_", fit_id, ".RData", sep = ''))
}

rhats <- bayesplot::rhat(fit)
message(sprintf("Maximum rhat is %f for male non-icp\n", max(rhats)))

original_Hb <- denormalize_vector(stan.preprocessed.male.nofix$y_test, stan.preprocessed.male.nofix$par_means["Hb"], stan.preprocessed.male.nofix$par_sds["Hb"]) 
scores <- get_scores(fit, params$Hb_cutoff_male, stan.preprocessed.male.nofix$par_means[["Hb"]], stan.preprocessed.male.nofix$par_sds[["Hb"]])
test_labels <- ifelse(original_Hb < params$Hb_cutoff_male, 1, 0)

val1.m <- validate_fit(fit, 
                       original_Hb, test_labels, params$Hb_cutoff_male, scores=scores,
                       pars1, vars1, metric = "mean", cat.plot = FALSE)
#val1.m

val1.m$posterior.plot
val1.m$scatter_plot
val1.m$conf.matrix.plot
val1.m$roc_plot
val1.m$pr_plot


error_rows[[length(error_rows)+1]] <- list(Model="Male, linear mixed model", MAE=val1.m$mae, RMSE=val1.m$rmse, 
                                           MAE2=val1.m$mae2, RMSE2=val1.m$rmse2, AUROC=val1.m$roc_auc, AUPR=val1.m$pr_auc)

save(val1.m, file = paste(rawresultdir, "raw_result_male_", fit_id, ".rdata", sep = ''))

# Remove unnescessary values to save memory
if (knit == TRUE) {rm(stan.lists.male.nofix, val1.m, fit); invisible(gc(reset = TRUE))}

```


```{asis, echo=compute_female_nofix}
### Model 1, female donors
```

```{r Model 1 female donors, results="hide", eval=compute_female_nofix}
if (!exists("stan.preprocessed.female.nofix"))
  load(paste(datadir,"stan_preprocessed_datasets_",data_id,".RData", sep = ''))
if (!exists("stan.lists.female.nofix"))
  load(paste(datadir,"stan_lists_female_nofix_",data_id,".RData", sep = ''))
warning("here1")
if (skip_train == TRUE) {
  load(paste(stanfitpath, "small_model1_female_", fit_id, ".RData", sep = ''))
} else {
  fit <- stan(file = stanf1, 
            data = stan.lists.female.nofix,
            save_warmup = TRUE,
            iter = params$iterations,
            control = list(max_treedepth = 12))
  save(fit, file = paste(stanfitpath, "small_model1_female_", fit_id, ".RData", sep = ''))
}

rhats <- rhat(fit)
message(sprintf("Maximum rhat is %f for female non-icp\n", max(rhats)))



original_Hb <- denormalize_vector(stan.preprocessed.female.nofix$y_test, stan.preprocessed.female.nofix$par_means["Hb"], stan.preprocessed.female.nofix$par_sds["Hb"]) 
scores <- get_scores(fit, params$Hb_cutoff_female, stan.preprocessed.female.nofix$par_means[["Hb"]], stan.preprocessed.female.nofix$par_sds[["Hb"]])
test_labels <- ifelse(original_Hb >= params$Hb_cutoff_female, 0, 1)

val1.f <- validate_fit(fit,  
                       original_Hb, test_labels, params$Hb_cutoff_female, scores,
                       pars1, vars1, metric = "mean", cat.plot = FALSE)

val1.f$posterior.plot
val1.f$scatter_plot
val1.f$conf.matrix.plot
val1.f$roc_plot
val1.f$pr_plot

error_rows[[length(error_rows)+1]] <- list(Model="Female, linear mixed model", MAE=val1.f$mae, RMSE=val1.f$rmse, 
                                        MAE2=val1.f$mae2, RMSE2=val1.f$rmse2, AUROC=val1.f$roc_auc, AUPR=val1.f$pr_auc)

save(val1.f, file = paste(rawresultdir, "raw_result_female_", fit_id, ".rdata", sep = ''))

if (knit == TRUE) {rm(stan.lists.female.nofix, val1.f, fit); invisible(gc(reset = TRUE))}

```


```{asis, echo=compute_male_icpfix}
### Model 1, male donors, icp-fix
```

```{r Model 1 male donors icp-fix, results="hide", eval=compute_male_icpfix}
message("heipparallaa")
if (!exists("stan.preprocessed.male.icpfix"))
  load(paste(datadir,"stan_preprocessed_datasets_",data_id,".RData", sep = ''))
if (!exists("stan.lists.male.icpfix"))
  load(paste(datadir,"stan_lists_male_icpfix_",data_id,".RData", sep = ''))
message("taalla ollaan")
if (skip_train == TRUE) {
  load(paste(stanfitpath, "small_model1_male_icp_", fit_id, ".RData", sep = ''))
} else {
  fit <- stan(file = stanf1.icp, 
            data = stan.lists.male.icpfix,
            save_warmup = TRUE,
            iter = params$iterations,
            control = list(max_treedepth = 12))
  save(fit, file = paste(stanfitpath, "small_model1_male_icp_", fit_id, ".RData", sep = ''))
}

rhats <- rhat(fit)
message(sprintf("Maximum rhat is %f for male icp\n", max(rhats)))



original_Hb <- denormalize_vector(stan.preprocessed.male.icpfix$y_test, stan.preprocessed.male.icpfix$par_means["Hb"], stan.preprocessed.male.icpfix$par_sds["Hb"]) 
scores <- get_scores(fit, params$Hb_cutoff_male, stan.preprocessed.male.icpfix$par_means[["Hb"]], stan.preprocessed.male.icpfix$par_sds[["Hb"]])
test_labels <- ifelse(original_Hb >= params$Hb_cutoff_male, 0, 1)

val1.m.icp <- validate_fit(fit, 
                           original_Hb, test_labels, params$Hb_cutoff_male, scores,
                           pars1.icp, vars1.icp, metric = "mean", cat.plot = FALSE)

val1.m.icp$posterior.plot
val1.m.icp$scatter_plot
val1.m.icp$conf.matrix.plot
val1.m.icp$roc_plot
val1.m.icp$pr_plot

error_rows[[length(error_rows)+1]] <- list(Model="Male, dynamic linear mixed model", MAE=val1.m.icp$mae, RMSE=val1.m.icp$rmse, 
                                        MAE2=val1.m.icp$mae2, RMSE2=val1.m.icp$rmse2, AUROC=val1.m.icp$roc_auc, AUPR=val1.m.icp$pr_auc)

save(val1.m.icp, file = paste(rawresultdir, "raw_result_male_icp_", fit_id, ".rdata", sep = ''))

if (knit == TRUE) {rm(stan.lists.male.icpfix, val1.m.icp, fit); invisible(gc(reset = TRUE))}

```

```{asis echo=compute_female_icpfix}
### Model 1, female donors, icp-fix
```

```{r Model 1 female donors icp-fix, results="hide", eval=compute_female_icpfix}
if (!exists("stan.preprocessed.female.icpfix"))
  load(paste(datadir,"stan_preprocessed_datasets_",data_id,".RData", sep = ''))
if (!exists("stan.lists.female.icpfix"))
  load(paste(datadir,"stan_lists_female_icpfix_",data_id,".RData", sep = ''))

if (skip_train == TRUE) {
  load(paste(stanfitpath, "small_model1_female_icp_", fit_id, ".RData", sep = ''))
} else {
  fit <- stan(file = stanf1.icp, 
            data = stan.lists.female.icpfix,
            save_warmup = TRUE,
            iter = params$iterations,
            control = list(max_treedepth = 12))
  save(fit, file = paste(stanfitpath, "small_model1_female_icp_", fit_id, ".RData", sep = ''))
}

rhats <- rhat(fit)
message(sprintf("Maximum rhat is %f for female icp\n", max(rhats)))



original_Hb <- denormalize_vector(stan.preprocessed.female.icpfix$y_test, stan.preprocessed.female.icpfix$par_means["Hb"], stan.preprocessed.female.icpfix$par_sds["Hb"]) 
scores <- get_scores(fit, params$Hb_cutoff_female, stan.preprocessed.female.icpfix$par_means[["Hb"]], stan.preprocessed.female.icpfix$par_sds[["Hb"]])
test_labels <- ifelse(original_Hb >= params$Hb_cutoff_female, 0, 1)

val1.f.icp <- validate_fit(fit, 
                           original_Hb, test_labels, params$Hb_cutoff_female, scores,
                           pars1.icp, vars1.icp, metric = "mean", cat.plot = FALSE)

val1.f.icp$posterior.plot
val1.f.icp$scatter_plot
val1.f.icp$conf.matrix.plot
val1.f.icp$roc_plot
val1.f.icp$pr_plot

error_rows[[length(error_rows)+1]] <- list(Model="Female, dynamic linear mixed model", MAE=val1.f.icp$mae, RMSE=val1.f.icp$rmse, 
                                        MAE2=val1.f.icp$mae2, RMSE2=val1.f.icp$rmse2, AUROC=val1.f.icp$roc_auc, AUPR=val1.f.icp$pr_auc)

save(val1.f.icp, file = paste(rawresultdir, "raw_result_female_icp_", fit_id, ".rdata", sep = ''))

if (knit == TRUE) {rm(stan.lists.female.icpfix, val1.f.icp, fit); invisible(gc(reset = TRUE))}

```




### Errors

```{r Errors}

#str(error_rows)
#df <- do.call(rbind, error_rows)
for (i in seq_along(error_rows)) {
  if (i == 1) {
    errors <- as_tibble(error_rows[[1]])
  } else {
    errors <- rbind(errors, error_rows[[i]])
  }
}  
#errors <- as_tibble(df)
errors <- errors %>% rename("MAE (g / L)" = MAE, "RMSE (g / L)" = RMSE,
                            "MAE (mmol / L)" = MAE2, "RMSE (mmol / L)" = RMSE2)

write_csv(errors, paste(rawresultdir, "errors_", fit_id, ".csv", sep = ''))
if (!is.null(params$errors_table_file)) {
  write_csv(errors, params$errors_table_file)   # Pass the table to the web app as well
}

```

```{r}
kable(errors, digits=3, format.args = list(digits=3))
```




